### PR TITLE
halo2_gadgets: Migrate chip gates to `Constraints::with_selector`

### PR DESCRIPTION
**Pull request by str4d**
Only one gate couldn't be migrated without altering the Orchard circuit.

### CODE REVIEW

**Comment by str4d** [comment]
Ignore the beta Clippy lints about `array::IntoInter::new`; they are because `beta` is 1.61, while we are currently pinned to MSRV 1.51 which doesn't have the APIs the beta lints refer to. The lints will be fixed once #482 is addressed.

File: halo2_gadgets/src/poseidon/pow5.rs [hunk 1]
```diff
@@ -94,20 +96,23 @@ impl<F: FieldExt, const WIDTH: usize, const RATE: usize> Pow5Chip<F, WIDTH, RATE
...
+                            .map(|idx| {
+                                let state_cur = meta.query_advice(state[idx], Rotation::cur());
+                                let rc_a = meta.query_fixed(rc_a[idx], Rotation::cur());
+                                pow_5(state_cur + rc_a) * m_reg[next_idx][idx]
+                            })
+                            .reduce(|acc, term| acc + term)
+                            .expect("WIDTH > 0");
+                        expr - state_next
+                    })
+                    .collect::<Vec<_>>(),
```
Comment (daira) [inline halo2_gadgets/src/poseidon/pow5.rs:114]: Collecting into a Vec shouldn't be necessary here, I think.
```suggestion
```

Comment (str4d) [reply halo2_gadgets/src/poseidon/pow5.rs:114]: It is necessary due to lifetimes. Believe me, I tried 🙂

**Review by daira — APPROVED** [review]
utACK

**Review by ebfull — APPROVED** [review]

**Pull request merged**